src/Setup/Data/Tool.purs

```diff
@@ -118,26 +119,31 @@ installMethod tool version = do
     Spago -> Tarball
       { source: formatGitHub'
           -- Spago has changed naming conventions from version to version
-          if version >= unsafeVersion "0.18.1" then case platform of
-            Windows -> "Windows"
-            Mac -> "macOS"
-            Linux -> "Linux"
-          else if version == unsafeVersion "0.18.0" then case platform of
-            Windows -> "windows-latest"
-            Mac -> "macOS-latest"
-            Linux -> "linux-latest"
-          else case platform of
-            Windows -> "windows"
-            Mac -> "osx"
-            Linux -> "linux"
+          case versionField of
+            Exact version 
+              | version >= unsafeVersion "0.18.1" -> 
+                case platform of
+                  Windows -> "Windows"
+                  Mac -> "macOS"
+                  Linux -> "Linux"
+              | version == unsafeVersion "0.18.1" -> 
+                case platform of
+                  Windows -> "windows-latest"
+                  Mac -> "macOS-latest"
+                  Linux -> "linux-latest"
+            _ -> 
+                case platform of
+                  Windows -> "windows"
+                  Mac -> "osx"
+                  Linux -> "linux"
```

Comment:
In this `Spago` case, we _know_ we've got an `Exact` version for `versionField`, but that's not represented in the types, so end up having to have a wildcard case. Struggling with how to "make impossible states unrepresentable" here...

Reply:
Yea, I think this is a sign that we haven't got the data structures quite right -- this feels awkward. Ideally, by the time we know we're installing a tarball we also know that we must have an exact version ready to go.